server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java

```diff
@@ -577,4 +590,96 @@ public IndexMetadataTransferException(String errorDesc, Throwable cause) {
             super(errorDesc, cause);
         }
     }
+
+    /**
+     * Deletes older than last {@code versionsToRetain} manifests. Also cleans up unreferenced IndexMetadata associated with older manifests
+     * @param clusterName name of the cluster
+     * @param clusterUUID uuid of cluster state to refer to in remote
+     * @param manifestsToRetain no of latest manifest files to keep in remote
+     */
+    public void deleteClusterMetadataMarker(String clusterName, String clusterUUID, int manifestsToRetain) {
+        BlobStoreTransferService transferService = new BlobStoreTransferService(blobStoreRepository.blobStore(), threadpool);
+
+        synchronized (this) {
+            transferService.listAllInSortedOrderAsync(
+                ThreadPool.Names.REMOTE_PURGE,
+                getManifestFolderPath(clusterName, clusterUUID),
+                "manifest",
+                Integer.MAX_VALUE,
+                new ActionListener<>() {
+                    int evaluatedManifestCount = 1;
+
+                    @Override
+                    public void onResponse(List<BlobMetadata> blobMetadata) {
+                        Set<String> filesToKeep = new HashSet<>();
+                        List<String> stalePaths = new ArrayList<>();
+                        blobMetadata.forEach(manifestBlobMetadata -> {
+                            ClusterMetadataManifest clusterMetadataManifest = fetchRemoteClusterMetadataManifest(
+                                clusterName,
+                                clusterUUID,
+                                manifestBlobMetadata.name()
+                            );
+                            if (evaluatedManifestCount <= manifestsToRetain) {
+                                clusterMetadataManifest.getIndices()
```

Comment:
Since we are adding manifest and index files in stale path together, can there be dangling index metadata file? A case where we failed to delete the index metadata but deleted the manifest.